[Bug]: Sandbox `.updateData` unexpected behavior regarding job data

### Version

v5.12.6

### Platform

NodeJS

### What happened?

I cannot get `job.updateData` and `job.moveToDelayed` methods working without having them fail the first couple of times. Then when the jobs are retried (like 2 or 3 times) it magically succeeds.

I am using Sandboxed workers of course 😄.

I'm trying to utilize the Step Pattern in your docs. But the only difference is: each step "builds on top of the jobData" of previous steps. But I keep losing the fields I've added in the step before. Its like `job.updateData()` updates the data in a "delayed" fashion if it contains new keys.

### How to reproduce.

Here's a reproduction within `bullmq` tests in my forked repo: https://github.com/taskforcesh/bullmq/compare/master...tolgap:bullmq:chore/reproduce-issue-2731?expand=1

### Relevant log output

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

@manast I agree with your sentiment. But here's a gist of my issue. I've tried to reproduce my issue in a branch here: https://github.com/taskforcesh/bullmq/compare/master...tolgap:bullmq:chore/reproduce-issue-2731?expand=1

I'm trying to utilize the Step Pattern in your docs. But the only difference is: each step "builds on top of the jobData" of previous steps. But I keep losing the fields I've added in the step before. Its like `job.updateData` updates the data in a "delayed" fashion if it contains new keys.

```console
$> yarn test


  sandbox step pattern
    1) should process steps and complete


  0 passing (546ms)
  1 failing

  1) sandbox step pattern
       should process steps and complete:

      AssertionError: expected { step: 'FINISH', …(1) } to deeply equal { step: 'FINISHED', …(2) }
      + expected - actual

       {
      -  "extraDataFinishedStep": "finished data"
      -  "step": "FINISH"
      +  "extraDataFinishedStep": "finish data"
      +  "extraDataForSecondStep": "second data"
      +  "step": "FINISHED"
       }
      
      at Worker.<anonymous> (tests/test_sandbox_step_pattern.ts:45:34)
      at Worker.emit (node:events:513:28)
      at Worker.emit (src/classes/queue-base.ts:111:20)
      at Worker.emit (src/classes/worker.ts:323:18)
      at handleCompleted (src/classes/worker.ts:757:14)
      at processTicksAndRejections (node:internal/process/task_queues:95:5)
      at Worker.processJob (src/classes/worker.ts:806:14)
      at Worker.retryIfFailed (src/classes/worker.ts:1007:16)
```

--- a/.husky/pre-commit
+++ b/.husky/pre-commit
@@ -1,4 +1,4 @@
 #!/usr/bin/env sh
 . "$(dirname -- "$0")/_/husky.sh"
 
-yarn npm-run-all pretty:quick lint:staged
+npx npm-run-all pretty:quick lint:staged

--- a/package.json
+++ b/package.json
@@ -28,7 +28,7 @@
     "copy:lua": "copyfiles -f ./src/commands/*.lua ./dist/cjs/commands && copyfiles -f ./src/commands/*.lua ./dist/esm/commands",
     "copy:lua:python": "copyfiles -f ./rawScripts/*.lua ./python/bullmq/commands",
     "copy:main:type": "copyfiles -f ./dist/esm/classes/main.d.ts ./dist/esm/classes/main-worker.d.ts ./dist/cjs/classes",
-    "coverage": "nyc --reporter=text --reporter=lcovonly yarn test",
+    "coverage": "nyc --reporter=text --reporter=lcovonly npm run test",
     "cm": "git cz",
     "docs": "typedoc --excludeExternals --excludeProtected --excludePrivate --readme none src/index.ts",
     "docs:json": "typedoc --excludeExternals --excludeProtected --excludePrivate --readme none src/index.ts --json ./apiVersions/v5.json --name v5",
@@ -39,7 +39,7 @@
     "generate:raw:scripts": "ts-node --project tsconfig-cjs.json generateRawScripts.ts",
     "lint": "./node_modules/.bin/eslint . --ignore-path ./.eslintignore",
     "lint:staged": "lint-staged",
-    "prepublishOnly": "yarn build",
+    "prepublishOnly": "npm run build",
     "prepare": "husky install",
     "pretest": "npm-run-all clean:scripts generate:raw:scripts transform:commands",
     "prettier": "prettier --config package.json src/**/*.ts",
@@ -162,7 +162,7 @@
     ]
   },
   "lint-staged": {
-    "*.{js,ts}": "yarn eslint:fix"
+    "*.{js,ts}": "npm run eslint:fix"
   },
   "repository": {
     "type": "git",

--- a/src/classes/child-processor.ts
+++ b/src/classes/child-processor.ts
@@ -114,7 +114,7 @@ export class ChildProcessor {
     job: JobJsonSandbox,
     send: (msg: any) => Promise<void>,
   ): SandboxedJob {
-    return {
+    const wrappedJob = {
       ...job,
       data: JSON.parse(job.data || '{}'),
       opts: job.opts,
@@ -136,7 +136,7 @@ export class ChildProcessor {
        * Emulate the real job `log` function.
        */
       log: async (row: any) => {
-        send({
+        await send({
           cmd: ParentCommand.Log,
           value: row,
         });
@@ -145,7 +145,7 @@ export class ChildProcessor {
        * Emulate the real job `moveToDelayed` function.
        */
       moveToDelayed: async (timestamp: number, token?: string) => {
-        send({
+        await send({
           cmd: ParentCommand.MoveToDelayed,
           value: { timestamp, token },
         });
@@ -154,11 +154,14 @@ export class ChildProcessor {
        * Emulate the real job `updateData` function.
        */
       updateData: async (data: any) => {
-        send({
+        await send({
           cmd: ParentCommand.Update,
           value: data,
         });
+        wrappedJob.data = data;
       },
     };
+
+    return wrappedJob;
   }
 }

--- a/tests/fixtures/fixture_processor_steps.js
+++ b/tests/fixtures/fixture_processor_steps.js
@@ -1,0 +1,37 @@
+'use strict';
+
+const delay = require('./delay');
+
+module.exports = async function (job) {
+  let step = job.data.step;
+  while (step !== 'FINISH') {
+    switch (step) {
+      case 'INITIAL': {
+        await delay(200);
+        const data = {
+          ...job.data,
+          step: 'SECOND',
+          extraDataSecondStep: 'second data',
+        };
+        await job.updateData(data);
+        step = 'SECOND';
+        break;
+      }
+      case 'SECOND': {
+        await delay(200);
+        const data = {
+          ...job.data,
+          extraDataFinishedStep: 'finish data',
+          step: 'FINISH',
+        };
+
+        await job.updateData(data);
+        step = 'FINISH';
+        return;
+      }
+      default: {
+        throw new Error('invalid step');
+      }
+    }
+  }
+};

--- a/tests/test_sandboxed_process.ts
+++ b/tests/test_sandboxed_process.ts
@@ -545,6 +545,37 @@ function sandboxProcessTests(
       await worker.close();
     });
 
+    it('should process steps and complete', async () => {
+      const processFile = __dirname + '/fixtures/fixture_processor_steps.js';
+
+      const worker = new Worker(queueName, processFile, {
+        connection,
+        prefix,
+        drainDelay: 1,
+      });
+
+      const completing = new Promise<void>((resolve, reject) => {
+        worker.on('completed', async (job: Job) => {
+          try {
+            expect(job.data).to.be.eql({
+              step: 'FINISH',
+              extraDataSecondStep: 'second data',
+              extraDataFinishedStep: 'finish data',
+            });
+            resolve();
+          } catch (err) {
+            reject(err);
+          }
+        });
+      });
+
+      await queue.add('test', { step: 'INITIAL' });
+
+      await completing;
+
+      await worker.close();
+    });
+
     it('should process and move to delayed', async () => {
       const processFile =
         __dirname + '/fixtures/fixture_processor_move_to_delayed.js';

